File: week1/pset1/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    //Prompt user about credit card number
    long number = get_long("Number: ");

    //Separate digits in credit card prompted

    //Find dividers 
    long divider[16];
    for (int i = 0; i < 16; i++)
    {
        divider[i] = (long) pow(10, (15 - i));
    }

    long numbers[16];
    for (int i = 0; i < 16; i++)
    {
        numbers[i] = number / divider[i];
        number = number % divider[i];
    }

    // Multiply digits
    int mnumber[16];
    for (int i = 0; i < 16; i++)
    {
        mnumber[i] = numbers[i] * 2;
    }

    // Separate digits from the doubles
    // Find tens from decimal
    int mnumberd [16];
    for (int i = 0; i < 16; i++)
    {
        mnumberd[i] = mnumber[i] / 10;
    }
    // Find unit from decimal
    int mnumberr [16];
    for (int i = 0; i < 16; i++)
    {
        mnumberr[i] = mnumber[i] % 10;
    }

    //Sum of tens of every other digit
    int sumtens = 0;
    for (int i = 0; i < 16; i = i + 2)
    {
        sumtens = sumtens + mnumberd[i];
    }

    //Sum of units of every other digit
    int sumunits = 0;
    for (int i = 0; i < 16; i = i + 2)
    {
        sumunits = sumunits + mnumberr[i];
    }

    //Sum of every other digit
    int sumdigit = 0;
    for (int i = 1; i < 16; i = i + 2)
    {
        sumdigit = sumdigit + numbers[i];
    }

    //Sum total of valid digits
    int sumtotal = sumtens + sumunits + sumdigit;

    //Check last digit of total
    int digit[3];
    for (int i = 0; i < 3; i++)
    {
        digit[i] = sumtotal / (int) pow(10, 2 - i);
        sumtotal = sumtotal % (int) pow(10, 2 - i);
    }

    //Check if credit card is valid
    if (digit[2] != 0)
    {
        printf("INVALID\n");
    }

    //Check if card is AMEX
    else if (numbers[0] == 0 && numbers[1] == 3)
    {
        if (numbers[2] == 4 || numbers[2] == 7)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

    //Check if card is MasterCard
    else if (numbers[0] == 5)
    {
        if (numbers[1] == 1 || numbers[1] == 2 || numbers[1] == 3 || numbers[1] == 4 || numbers[1] == 5)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

    //Check if card is Visa
    else if (numbers[0] == 4)
    {
        printf("VISA\n");
    }
    else if (numbers[0] == 0 && numbers[1] == 0 && numbers[2] == 0 && numbers[3] == 4)
    {
        printf("VISA\n");
    }

    //Everything else is invalid
    else
    {
        printf("INVALID\n");
    }
}

File: week6/pset6/credit.py
import cs50
import sys


def main():
    # Get input from user
    cc_number = getnumber()
    # Check Luhn's Algorithm
    sumdigits = luhn(cc_number)
    if sumdigits % 10 == 0:
        # Check if AMEX
        if len(cc_number) == 15:
            if cc_number[0] == "3":
                if cc_number[1] == "4" or cc_number[1] == "7":
                    print("AMEX")
                    sys.exit(0)
        # Check if MasterCard
        if len(cc_number) == 16:
            if cc_number[0] == "5":
                if cc_number[1] in ["1", "2", "3", "4", "5"]:
                    print("MASTERCARD")
                    sys.exit(0)
        # Check if Visa
        if len(cc_number) in [13, 14, 15, 16]:
            if cc_number[0] == "4":
                print("VISA")
                sys.exit(0)
    print("INVALID")
    sys.exit(1)


# Check is user prompted a number
def getnumber():
    while True:
        cc_number = cs50.get_string("Number: ")
        if cc_number.isdigit():
            break
    return cc_number


# Check for Luhn's Algorithm
def luhn(cc_number):
    sumdigits = 0
    for i in range(len(cc_number)):
        if len(cc_number) % 2 == 0:
            if i % 2 == 0:
                product = (int(cc_number[i]) * 2)
                sumdigits += int(product / 10) + product % 10
            else:
                sumdigits += int(cc_number[i])
        else:
            if i % 2 == 1:
                product = (int(cc_number[i]) * 2)
                sumdigits += int(product / 10) + product % 10
            else:
                sumdigits += int(cc_number[i])
    return sumdigits


if __name__ == "__main__":
    main()

